Panic when running `istioctl manifest translate` subcommand: runtime error: index out of range

### Is this the right place to submit this?

- [x] This is not a security vulnerability or a crashing bug
- [x] This is not a question about how to use Istio

### Bug Description

Hi, I'm in the process of upgrading Istio to version 1.25.2 and as I previously used the Istio Operator for the install, I need to migrate to using Helm. As per the [deprecation notice](https://istio.io/latest/blog/2024/in-cluster-operator-deprecation-announcement/#migrating-to-helm), I'm trying to use the `istioctl manifest translate` subcommand to translate my IstioOperator CR to Helm values.

First I got the IstioOperator resource from the cluster (and removed the .status field):

```kubectl get istiooperators -n service-mesh-operator istio-operator -o yaml > istio-operator-redacted.yaml```

When running the command with the attached [istio-operator-redacted.yaml.txt](https://github.com/user-attachments/files/20105119/istio-operator-redacted.yaml.txt) file, I get the following panic and error message.

```
istioctl_1.25.2 manifest translate -f istio-operator-redacted.yaml -o helm
panic: runtime error: index out of range [5] with length 5

goroutine 1 [running]:
istio.io/istio/operator/cmd/mesh.ManifestTranslate({0x0, 0x0}, 0x14000987e00, {0x107a7eab8, 0x14000a3fda0})
	istio.io/istio/operator/cmd/mesh/manifest-translate.go:162 +0x1214
istio.io/istio/operator/cmd/mesh.ManifestCmd.ManifestTranslateCmd.func4(0x14000ae8908, {0x1400090bb80?, 0x4?, 0x1069cad8f?})
	istio.io/istio/operator/cmd/mesh/manifest-translate.go:102 +0x140
github.com/spf13/cobra.(*Command).execute(0x14000ae8908, {0x1400090bb40, 0x4, 0x4})
	github.com/spf13/cobra@v1.8.1/command.go:985 +0x834
github.com/spf13/cobra.(*Command).ExecuteC(0x140003d8908)
	github.com/spf13/cobra@v1.8.1/command.go:1117 +0x344
github.com/spf13/cobra.(*Command).Execute(0x140001de1d0?)
	github.com/spf13/cobra@v1.8.1/command.go:1041 +0x1c
main.main()
	istio.io/istio/istioctl/cmd/istioctl/main.go:39 +0xd0
```

Despite this, I still seem to get some of the expected output files (no README.md). Weirdly though, the `base-values.yaml`, `cni-values.yaml` and `pilot-values.yaml`, are all exactly the same and the values within them do not seem to correspond to the values [defined in their respective charts](https://artifacthub.io/packages/helm/istio-official/istiod/1.25.0?modal=values) which I think they should do?

I have found that removing certain parts of the IstioOperator CR cause the issue to go away, but it seems a bit random, for instance, setting `.spec.components.cni.enabled` to `false`, or setting `.spec.components.ingressgateways[].enabled` to `false` for either istio-ingressgateway or ingressgateway-internal-routing both make the panic go away.

### Version

```Text
istioctl_1.25.2 version
Istio is not present in the cluster: no running Istio pods in namespace "istio-system"
client version: 1.25.2
```

### Additional Information

My IstioOperator YAML file was pulled directly from the cluster and is definitely all valid:

```
istioctl_1.25.2 validate -f istio-operator-redacted.yaml
"istio-operator-redacted.yaml" is valid
```

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Dual Stack
- [x] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [x] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

--- a/operator/cmd/mesh/manifest-translate.go
+++ b/operator/cmd/mesh/manifest-translate.go
@@ -18,8 +18,6 @@ import (
 	_ "embed"
 	"fmt"
 	"io/fs"
-	"istio.io/istio/operator/pkg/component"
-	"istio.io/istio/operator/pkg/manifest"
 	"os"
 	"path/filepath"
 	"strings"
@@ -27,6 +25,8 @@ import (
 	"github.com/spf13/cobra"
 
 	"istio.io/istio/istioctl/pkg/cli"
+	"istio.io/istio/operator/pkg/component"
+	"istio.io/istio/operator/pkg/manifest"
 	"istio.io/istio/operator/pkg/render"
 	"istio.io/istio/operator/pkg/util/clog"
 	"istio.io/istio/pkg/kube"

--- a/operator/cmd/mesh/manifest-translate.go
+++ b/operator/cmd/mesh/manifest-translate.go
@@ -18,6 +18,8 @@ import (
 	_ "embed"
 	"fmt"
 	"io/fs"
+	"istio.io/istio/operator/pkg/component"
+	"istio.io/istio/operator/pkg/manifest"
 	"os"
 	"path/filepath"
 	"strings"
@@ -110,11 +112,14 @@ func ManifestTranslate(kubeClient kube.CLIClient, mgArgs *ManifestTranslateArgs,
 	if err != nil {
 		return err
 	}
+	generatedManifestMap := make(map[component.Name]manifest.ManifestSet)
+	for _, m := range istioctlGeneratedManifests {
+		generatedManifestMap[m.Component] = m
+	}
 	res, err := render.Migrate(mgArgs.InFilenames, setFlags, kubeClient)
 	if err != nil {
 		return err
 	}
-	_ = res
 	out := mgArgs.Output
 	write := func(name string, contents string) error {
 		perm := 0o644
@@ -124,7 +129,7 @@ func ManifestTranslate(kubeClient kube.CLIClient, mgArgs *ManifestTranslateArgs,
 		return os.WriteFile(filepath.Join(out, name), []byte(contents), fs.FileMode(perm))
 	}
 	results := []string{}
-	for idx, info := range res.Components {
+	for _, info := range res.Components {
 		name := ptr.NonEmptyOrDefault(info.ComponentSpec.Name, info.Component.SpecName)
 		if info.Component.ReleaseName == "" {
 			results = append(results, fmt.Sprintf(`* ❌ **Component %s**: migration is **NOT** directly supported!`,
@@ -158,7 +163,11 @@ func ManifestTranslate(kubeClient kube.CLIClient, mgArgs *ManifestTranslateArgs,
 		}
 		diffWarn := ""
 		helmManifests := strings.Join(sortManifests(info.Manifest), "\n---\n")
-		istioctlManifests := strings.Join(sortManifests(istioctlGeneratedManifests[idx].Manifests), "\n---\n")
+		generatedManifest, ok := generatedManifestMap[info.Component.UserFacingName]
+		if !ok {
+			continue
+		}
+		istioctlManifests := strings.Join(sortManifests(generatedManifest.Manifests), "\n---\n")
 		if helmManifests != istioctlManifests {
 			helmName := fmt.Sprintf("diff-%s-helm-output.yaml", name)
 			istioctlName := fmt.Sprintf("diff-%s-istioctl-output.yaml", name)

--- a/releasenotes/notes/56240.yaml
+++ b/releasenotes/notes/56240.yaml
@@ -4,4 +4,4 @@ area: installation
 issue: [56223]
 releaseNotes:
   - |
-    **Fixed** an issue that istioctl panic when running `manifest translate` with multiple gateways in IstioOperator configuration file.
+    **Fixed** a panic in `istioctl manifest translate` when the IstioOperator config contains multiple gateways.

--- a/releasenotes/notes/56240.yaml
+++ b/releasenotes/notes/56240.yaml
@@ -1,0 +1,7 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: installation
+issue: [56223]
+releaseNotes:
+  - |
+    **Fixed** an issue that istioctl panic when running `manifest translate` with multiple gateways in IstioOperator configuration file.

